### PR TITLE
Uninstall changes + LSO 

### PR DESCRIPTION
**Pull request by annagitel**
Signed-off-by: annagitel <asandler@redhat.com>

### CODE REVIEW

File: ocs_ci/ocs/uninstall.py [hunk 1]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+def uninstall_lso():
+    """
+    Function uninstalls local-volume objects from OCS cluster
+
+    """
+    ocp_obj = ocp.OCP()
+    storage_cluster = ocp.OCP(kind=constants.STORAGECLUSTER, resource_name=constants.DEFAULT_CLUSTERNAME)
+
+    sc_name = storage_cluster.get('spec').get('storageDeviceSets') \
+        .get().get('dataPVCTemplate').get('spec').get('storageClassName')
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:67]:
```python
    sc_name = storage_cluster.get('spec').get('storageDeviceSets') 
    .get().get('dataPVCTemplate').get('spec').get('storageClassName')
```

File: ocs_ci/ocs/uninstall.py [hunk 2]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+    sc_obj = ocp.OCP(kind=constants.STORAGECLASS, resource_name=sc_name)
+
+    lv_name = sc_obj.get('metadata').get('labels').get('local.storage.openshift.io/owner-name')
+    lv_obj = ocp.OCP(kind=constants.LOCAL_VOLUME, resource_name=lv_name)
+    log.info(f"storage class: {sc_name}  local volume:{lv_name}")
+    device_list = lv_obj.get('spec').get('storageClassDevices').get().get('devicePaths')
+    storage_node_list = get_labeled_nodes(constants.OPERATOR_NODE_LABEL)
+
+    log.info("deleting local volume")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:76]: please log the local volume details (name/path)

File: ocs_ci/ocs/uninstall.py [hunk 3]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+    lv_obj.delete()
+
+    pv_obj_list = ocp.OCP(kind=constants.PV, namespace=constants.LOCAL_STORAGE_NAMESPACE,
+                          selector=f"storage.openshift.com/local-volume-owner-name={sc_name}")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:80]:
```python
    pv_obj_list = (
        ocp.OCP(
            kind=constants.PV, namespace=constants.LOCAL_STORAGE_NAMESPACE,
            selector=f"storage.openshift.com/local-volume-owner-name={sc_name}"
        )
    )
```

File: ocs_ci/ocs/uninstall.py [hunk 4]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+    log.info("deleting local volume PVs")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:81]: please log the local volume details (name/path)

File: ocs_ci/ocs/uninstall.py [hunk 5]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+    for pv in pv_obj_list.get():
+        log.info(f"deleting pv {pv.get('metadata').get('name')}")
+        pv_obj_list.delete(resource_name=pv.get('metadata').get('name'))
+
+    log.info("deleting storage class")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:86]: please log the storage class name

File: ocs_ci/ocs/uninstall.py [hunk 6]
```diff
@@ -54,6 +55,55 @@ def remove_ocp_registry_from_ocs(platform):
...
+    sc_obj.delete()
+
+    log.info("removing local volume from storage nodes")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:89]: same

File: ocs_ci/ocs/uninstall.py (hunk 4 above)
Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:87]: done in every iteration

File: ocs_ci/ocs/uninstall.py (hunk 6 above)
Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:93]: dome in every iteration

File: ocs_ci/ocs/uninstall.py [hunk 7]
```diff
@@ -54,6 +55,68 @@ def remove_ocp_registry_from_ocs(platform):
         log.info("platform registry not supported")
 
 
+def uninstall_lso():
+    """
+    Function uninstalls local-volume objects from OCS cluster
+
+    """
+    ocp_obj = ocp.OCP()
+    storage_cluster = ocp.OCP(kind=constants.STORAGECLUSTER,
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:64]: To keep consistency, please change to:
```python
storage_cluster = (
    ocp.OCP(
        kind=constants.STORAGECLUSTER, 
        resource_name=constants.DEFAULT_CLUSTERNAME,
        namespace='openshift-storage'
    )
)
```

File: ocs_ci/ocs/uninstall.py [hunk 8]
```diff
@@ -54,6 +55,68 @@ def remove_ocp_registry_from_ocs(platform):
...
+                              resource_name=constants.DEFAULT_CLUSTERNAME,
+                              namespace='openshift-storage')
+
+    sc_name = storage_cluster.get().get('spec').get(
+        'storageDeviceSets')[0].get('dataPVCTemplate').get('spec').get('storageClassName')
+    sc_obj = ocp.OCP(kind=constants.STORAGECLASS,
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:70]: same

File: ocs_ci/ocs/uninstall.py [hunk 9]
```diff
@@ -54,6 +55,68 @@ def remove_ocp_registry_from_ocs(platform):
...
+                     resource_name=sc_name,
+                     namespace=constants.LOCAL_STORAGE_NAMESPACE
+                     )
+
+    lv_name = sc_obj.get().get('metadata').get('labels').get('local.storage.openshift.io/owner-name')
+    lv_obj = ocp.OCP(kind=constants.LOCAL_VOLUME,
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:76]: same

File: ocs_ci/ocs/uninstall.py [hunk 10]
```diff
@@ -54,6 +55,68 @@ def remove_ocp_registry_from_ocs(platform):
...
+                     resource_name=lv_name,
+                     namespace=constants.LOCAL_STORAGE_NAMESPACE
+                     )
+
+    log.info(f"storage class: {sc_name}  local volume:{lv_name}")
+
+    device_list = lv_obj.get().get('spec').get('storageClassDevices')[0].get('devicePaths')
+    storage_node_list = get_labeled_nodes(constants.OPERATOR_NODE_LABEL)
+
+    pv_obj_list = ocp.OCP(kind=constants.PV,
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:86]: same

File: ocs_ci/ocs/uninstall.py [hunk 11]
```diff
@@ -54,6 +55,68 @@ def remove_ocp_registry_from_ocs(platform):
...
+    Function uninstalls local-volume objects from OCS cluster
+
+    """
+    ocp_obj = ocp.OCP()
+    storage_cluster = ocp.OCP(kind=constants.STORAGECLUSTER,
+                              resource_name=constants.DEFAULT_CLUSTERNAME,
+                              namespace='openshift-storage')
+
+    sc_name = storage_cluster.get().get('spec').get(
+        'storageDeviceSets')[0].get('dataPVCTemplate').get('spec').get('storageClassName')
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:69]:
```python
sc_name = (
    storage_cluster.get().get('spec').get('storageDeviceSets')[0]
    .get('dataPVCTemplate').get('spec').get('storageClassName')
)
```

Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:69]: then moving the . get one line down it adds \, then deleting it, error is marked

File: ocs_ci/ocs/uninstall.py (hunk 7 above)
Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py:64]: not fixed

File: ocs_ci/ocs/uninstall.py (hunk 11 above)
Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py:69]: not fixed

File: ocs_ci/ocs/uninstall.py (hunk 7 above)
Comment (annagitel) [reply ocs_ci/ocs/uninstall.py]: forgot to push :)

File: ocs_ci/ocs/uninstall.py [hunk 12]
```diff
@@ -37,23 +36,96 @@ def remove_ocp_registry_from_ocs(platform):
...
+    if param_cmd1:
         image_registry_obj.patch(
-            resource_name=constants.IMAGE_REGISTRY_RESOURCE_NAME, params=param_cmd, format_type='json'
+            params=param_cmd1, format_type='json'
+        )
+        image_registry_obj.patch(
+            params=param_cmd2, format_type='json'
         )
     else:
         log.info("platform registry not supported")
```
Comment (ebenahar) [inline ocs_ci/ocs/uninstall.py:59]:
```python
        params_list = list()
        if platform.lower() == constants.AWS_PLATFORM:
            params_list.apend('[{"op": "remove", "path": "/spec/storage"}]')
            params_list.apend('[{"op": "remove", "path": "/status/storage"}]')

        elif platform.lower() == constants.VSPHERE_PLATFORM:
            params_list.apend('[{"op": "replace", "path": "/spec/storage", "value": {"emptyDir": "{}"}}]')
            params_list.apend('[{"op": "replace", "path": "/status/storage", "value": {"emptyDir": "{}"}}]')

        if params_list:
            for params in params_list:
                image_registry_obj.patch(params=params, format_type='json')
        else:
            log.info("platform registry not supported")
```

Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py:59]: BTW, what about other platforms?

Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:59]: new docs we have is only for azure. will implement with the 4.5 changes in a followup pr

Comment (prsurve) [reply ocs_ci/ocs/uninstall.py:59]: BM Platform missing  here

Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:59]: @prsurve do you have the BM docs? I will add this platform too

Comment (prsurve) [reply ocs_ci/ocs/uninstall.py:59]: Vmware LSO changes will  be applicable for BM LSO

Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py:60]: @annagitel above comment is not addressed

**Comment by annagitel** [comment]
[default-deployment.txt](https://github.com/red-hat-storage/ocs-ci/files/5091108/default-deployment.txt)
[lso-deployment.txt](https://github.com/red-hat-storage/ocs-ci/files/5091110/lso-deployment.txt)
verification logs

File: ocs_ci/ocs/uninstall.py (hunk 12 above)
Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py]: @annagitel my initial comment above is still not updated

Comment (annagitel) [reply ocs_ci/ocs/uninstall.py]: firs comment changes done, for the other platforms - will be implemented in a followup PR

Comment (ebenahar) [reply ocs_ci/ocs/uninstall.py]: thanks :+1:

File: ocs_ci/ocs/uninstall.py [hunk 13]
```diff
@@ -37,23 +37,112 @@ def remove_ocp_registry_from_ocs(platform):
...
+def remove_cluster_logging_operator_from_ocs():
+    """
+    Function removes cluster logging operator from OCS
+
+    """
+    # Deleting the clusterlogging instance
+    clusterlogging_obj = ocp.OCP(
+        kind=constants.CLUSTER_LOGGING, namespace=constants.OPENSHIFT_LOGGING_NAMESPACE
+    )
+    assert clusterlogging_obj.delete(resource_name='instance')
```
[the remark this answers is not in the document]
Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:67]: delete method hat a 'wait' param set to True by default

File: ocs_ci/ocs/uninstall.py [hunk 14]
```diff
@@ -37,23 +37,112 @@ def remove_ocp_registry_from_ocs(platform):
...
+
+    # Deleting the PVCs
+    pvc_obj = ocp.OCP(
+        kind=constants.PVC, namespace=constants.OPENSHIFT_LOGGING_NAMESPACE
+    )
+    pvc_list = get_all_pvcs(namespace=constants.OPENSHIFT_LOGGING_NAMESPACE)
+    for pvc in range(len(pvc_list) - 1):
+        pvc_obj.delete(resource_name=pvc_list['items'][pvc]['metadata']['name'])
+        pvc_obj.wait_for_delete(resource_name=pvc_list['items'][pvc]['metadata']['name'])
```
Comment (RazTamir) [inline ocs_ci/ocs/uninstall.py:76]: can we cycle on the PVCs for waiting for deletion after we initiate deletion for all of them?

File: ocs_ci/ocs/uninstall.py [hunk 15]
```diff
@@ -37,23 +37,112 @@ def remove_ocp_registry_from_ocs(platform):
...
+    lv_name = sc_obj.get().get('metadata').get('labels').get('local.storage.openshift.io/owner-name')
+    lv_obj = (
+        ocp.OCP(
+            kind=constants.LOCAL_VOLUME,
+            resource_name=lv_name,
+            namespace=constants.LOCAL_STORAGE_NAMESPACE
+        )
+    )
+
+    log.info(f"storage class: {lso_sc}  local volume:{lv_name}")
```
Comment (RazTamir) [inline ocs_ci/ocs/uninstall.py:103]: Please add more informative logging message

File: ocs_ci/ocs/uninstall.py (hunk 14 above)
Comment (annagitel) [reply ocs_ci/ocs/uninstall.py:76]: I think this wait can be removed anyway - delete has a wait param set to wait by default

File: ocs_ci/ocs/uninstall.py [hunk 16]
```diff
@@ -37,23 +37,112 @@ def remove_ocp_registry_from_ocs(platform):
...
+            namespace=constants.LOCAL_STORAGE_NAMESPACE
+        )
+    )
+
+    log.info("deleting local volume PVs")
+    for pv in pv_obj_list.get().get('items'):
+        log.info(f"deleting pv {pv.get('metadata').get('name')}")
+        pv_obj_list.delete(resource_name=pv.get('metadata').get('name'))
+
+    log.info("removing local volume from storage nodes")
```
Comment (RazTamir) [inline ocs_ci/ocs/uninstall.py:121]: Please use capital letters.
*R*emoving local ...

File: ocs_ci/ocs/uninstall.py [hunk 17]
```diff
@@ -64,39 +153,13 @@ def uninstall_ocs():
...
-    for pod_obj in all_pods:
-        try:
-            pvc_name = get_pvc_name(pod_obj)
-        except UnavailableResourceException:
-            continue
-        if pvc_name in pvc_name_list:
-            pods_to_delete.append(pod_obj)
+    for sc in sc_list:
+        pvc_to_delete.extend(pvc for pvc in get_all_pvcs_in_storageclass(
+            sc.get('metadata').get('name')) if 'noobaa' not in pvc.name)
```
Comment (RazTamir) [inline ocs_ci/ocs/uninstall.py:162]: close ) in new line

**Review by ebenahar — APPROVED** [review]

File: ocs_ci/ocs/uninstall.py [hunk 18]
```diff
@@ -105,59 +168,67 @@ def uninstall_ocs():
...
+
     log.info("Delete the storage classes with an openshift-storage provisioner list")
     for storage_class in sc_list:
-        log.info(f"deleting storage class {storage_class.get('metadata').get('name')}")
+        log.info(f"Deleting storage class {storage_class.get('metadata').get('name')}")
         sc_obj = ocp.OCP(kind=constants.STORAGECLASS)
         sc_obj.delete(resource_name=storage_class.get('metadata').get('name'))
 
-    log.info("unlabaling storage nodes")
+    log.info("Unlabaling storage nodes")
```
Comment (fbalak) [inline ocs_ci/ocs/uninstall.py:227]:
```suggestion
    log.info("Unlabeling storage nodes")
```

**Review by fbalak — APPROVED** [review]

File: ocs_ci/ocs/uninstall.py [hunk 19]
```diff
@@ -105,59 +168,67 @@ def uninstall_ocs():
...
-    logging_csv = csv.get().get('items')
-    if logging_csv:
-        clusterlogging_obj = ocp.OCP(
-            kind=constants.CLUSTER_LOGGING, namespace=constants.OPENSHIFT_LOGGING_NAMESPACE
-        )
-        clusterlogging_obj.delete(resource_name='instance')
+    try:
+        remove_cluster_logging_operator_from_ocs()
+    except CommandFailed:
+        log.info("no cluster logging found")
```
Comment (jimdevops19) [inline ocs_ci/ocs/uninstall.py:174]:
```suggestion
        log.info("No cluster logging found")
```

**Review by jimdevops19 — APPROVED** [review]

**Review by ebenahar — APPROVED** [review]

**Review by fbalak — APPROVED** [review]

**Pull request merged**